### PR TITLE
Add Simplex::FE_P and Simplex::FE_DGP

### PR DESCRIPTION
**Pull request by peterrum**
This PR adds a first version of `Simplex::FE_P` ~and `Simplex::MappingP`. I have put both into the same PR because currently latter uses `Simplex::FE_P`.~ Since I am not in expert in `FiniteElement` and `Mapping`, any help is welcome!

This PR depends on a couple of PRs. I have integrated  the changes there also here (so that the code compiles and) to make cooperation simper.  The relevant files are:
- include/deal.II/simplex/fe_p.h
- ~include/deal.II/simplex/mapping_p.h~
- source/simplex/fe_p.cc

### CODE REVIEW

**Comment by drwells** [comment]
Awesome! I will dig into this a little later today.

File: include/deal.II/grid/reference_cell.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include <deal.II/base/config.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * A namespace for reference cells.
+ */
+namespace RefernceCell
```
Comment (tjhei) [inline include/deal.II/grid/reference_cell.h:28]: typo

File: include/deal.II/grid/reference_cell.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    Wedge   = 6,
+    Hex     = 7,
+    Invalid = static_cast<std::uint8_t>(-1)
+  };
+} // namespace RefernceCell
+
+
+DEAL_II_NAMESPACE_CLOSE
+
+#endif
```
Comment (tjhei) [inline include/deal.II/grid/reference_cell.h:50]: missing newline

File: include/deal.II/simplex/fe_p.h [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+          cell_similarity != CellSimilarity::translation)
+        {
+          for (unsigned int k = 0; k < dofs_per_cell; ++k)
+            mapping.transform(make_array_view(fe_data.shape_3rd_derivatives, k),
+                              mapping_covariant_hessian,
+                              mapping_internal,
+                              make_array_view(output_data.shape_3rd_derivatives,
+                                              k));
+
+          // for (unsigned int k = 0; k < this->dofs_per_cell; ++k)
```
Comment (tjhei) [inline include/deal.II/simplex/fe_p.h:262]: remove or add assert?

File: include/deal.II/simplex/fe_p.h [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+       */
+      Table<2, Tensor<3, dim>> shape_3rd_derivatives;
+    };
+  };
+
+} // namespace Simplex
+
+DEAL_II_NAMESPACE_CLOSE
+
+#endif
```
Comment (tjhei) [inline include/deal.II/simplex/fe_p.h:324]: missing newline

File: include/deal.II/simplex/polynomials.h [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    for (unsigned int i = 0; i < dim; i++)
+      der[i] = grad[i];
+
+    return der;
+  }
+} // namespace Simplex
+
+DEAL_II_NAMESPACE_CLOSE
+
+#endif
```
Comment (tjhei) [inline include/deal.II/simplex/polynomials.h:167]: you seem to miss a newline in every other file. :-)

File: source/simplex/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+SET(_inst
+  fe_p.inst.in
+)
+
+FILE(GLOB _header
+  ${CMAKE_SOURCE_DIR}/include/deal.II/tet/*.h
+  )
+
+DEAL_II_ADD_LIBRARY(obj_tet OBJECT ${_src} ${_header} ${_inst})
```
Comment (tjhei) [inline source/simplex/CMakeLists.txt:43]: as in the other PR, this should be obj_simplex.

File: source/simplex/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+EXPAND_INSTANTIATIONS(obj_tet "${_inst}")
```
Comment (tjhei) [inline source/simplex/CMakeLists.txt:44]: :-)

File: source/simplex/fe_p.cc [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+            this->unit_support_points.emplace_back(0.0, 1.0, 0.0);
+            this->unit_support_points.emplace_back(0.0, 0.0, 1.0);
+            this->unit_support_points.emplace_back(0.5, 0.0, 0.0);
+            this->unit_support_points.emplace_back(0.5, 0.5, 0.0);
+            this->unit_support_points.emplace_back(0.0, 0.5, 0.0);
+            this->unit_support_points.emplace_back(0.0, 0.0, 0.5);
+            this->unit_support_points.emplace_back(0.5, 0.0, 0.5);
+            this->unit_support_points.emplace_back(0.0, 0.5, 0.5);
+
+            // TODO
```
Comment (tjhei) [inline source/simplex/fe_p.cc:114]: ?

**Comment by peterrum** [comment]
I have remove the mapping from this PR. I am working on an alternative more generic version and will create a separate PR.

**Comment by drwells** [comment]
The failing quicktest may be relevant:
```
An error occurred in line <1014> of file </jenkins/workspace/dealii_PR-10646/include/deal.II/fe/fe_tools.templates.h> in function

    void dealii::FETools::Compositing::build_face_tables(std::vector<std::pair<std::pair<unsigned int, unsigned int>, unsigned int>
>&, std::vector<std::pair<unsigned int, unsigned int> >&, const dealii::FiniteElement<dim, spacedim>&, bool) [with int dim = 2; 
int spacedim = 2]

The violated condition was: 

    total_index == fe.dofs_per_face

[...]

Stacktrace:

-----------

#0  /home/dealii/build/lib/libdeal_II.g.so.9.3.0-pre: void dealii::FETools::Compositing::build_face_tables<2, 
2>(std::vector<std::pair<std::pair<unsigned int, unsigned int>, unsigned int>, std::allocator<std::pair<std::pair<unsigned int, 
unsigned int>, unsigned int> > >&, std::vector<std::pair<unsigned int, unsigned int>, std::allocator<std::pair<unsigned int, 
unsigned int> > >&, dealii::FiniteElement<2, 2> const&, bool)

#1  /home/dealii/build/lib/libdeal_II.g.so.9.3.0-pre: dealii::FESystem<2, 2>::initialize(std::vector<dealii::FiniteElement<2, 2> const*,
 std::allocator<dealii::FiniteElement<2, 2> const*> > const&, std::vector<unsigned int, std::allocator<unsigned int> > const&)

#2  /home/dealii/build/lib/libdeal_II.g.so.9.3.0-pre: dealii::FESystem<2, 2>::FESystem(dealii::FiniteElement<2, 2> const&, unsigned 
int, dealii::FiniteElement<2, 2> const&, unsigned int)

#3  ./p4est.debug: void test<2>()

#4  ./p4est.debug: main
```

**Comment by peterrum** [comment]
> The failing quicktest may be relevant: 

This is already fixed in #10645. I will rebase in a moment.

**Comment by peterrum** [comment]
I have restructured the code such that now there is a base class from which `FE_P` and `FE_DGP` inherit. Once PR #10645 is merged, the changes to `FiniteElementData` is not part of this PR anymore.

Furthermore, I have added two tests.

**Pull request merged**